splice_machine/src/test/java/com/splicemachine/benchmark/FeatureStoreTriggerBenchmark.java

```diff
@@ -0,0 +1,498 @@
+        taskId.set(0);
+        runBenchmark(connections, () -> populateStagingTable(curTimestamp, nextTimestamp));
+        curTimestamp = nextTimestamp;
+    }
+
+    public void insertAndUpdateFromBatches() {
+        try (Connection conn = makeConnection()) {
+            try (Statement statement = conn.createStatement()) {
+                for (int update = 1; update <= updates; ++update) {
+                    long start = System.currentTimeMillis();
+                    int count = statement.executeUpdate(
+                        String.format("INSERT INTO FeatureTable --splice-properties insertMode=UPSERT\n" +
+                                        "SELECT entity_key, ts, feature1, feature2 " +
+                                        "FROM FeatureStaging WHERE update_order=%d", update)
+                    );
+                    long end = System.currentTimeMillis();
+                    updateStats(STAT_UPDATE, count, end - start);
+                }
+            }
+        }
+        catch (Throwable t) {
+            LOG.error("Connection broken", t);
+            abort = true;
+        }
+    }
+
+    private void createFeatureTableTriggers() throws SQLException {
+        testStatement.execute(
+                "CREATE TRIGGER FeatureTable_INSERTS " +
+                        "AFTER INSERT ON FeatureTable " +
+                        "REFERENCING NEW_TABLE AS NEWW " +
+                        "FOR EACH STATEMENT " +
+                        "INSERT INTO FeatureTableHistory (entity_key, asof_ts, ingest_ts, feature1, feature2) --splice-properties insertMode=UPSERT\n" +
+                        "SELECT NEWW.entity_key, NEWW.last_update_ts, CURRENT_TIMESTAMP, NEWW.feature1, NEWW.feature2 FROM NEWW"
+        );
+        testStatement.execute(
+                "CREATE TRIGGER FeatureTable_UPDATES " +
+                        "AFTER UPDATE ON FeatureTable " +
+                        "REFERENCING NEW_TABLE AS NEWW " +
+                        "FOR EACH STATEMENT " +
+                        "INSERT INTO FeatureTableHistory ( entity_key, asof_ts, ingest_ts, feature1, feature2) --splice-properties insertMode=UPSERT\n" +
+                        "SELECT NEWW.entity_key, NEWW.last_update_ts, CURRENT_TIMESTAMP, NEWW.feature1, NEWW.feature2 FROM NEWW"
+        );
+        testStatement.execute(
+                "CREATE TRIGGER FeatureTableHistory_UPDATE_OOO " +
+                        "AFTER UPDATE ON FeatureTable " +
+                        "REFERENCING NEW_TABLE AS NEWW OLD_TABLE AS OLDW " +
+                        "FOR EACH STATEMENT " +
+                        "INSERT INTO FeatureTable (entity_key, last_update_ts, feature1, feature2) --splice-properties insertMode=UPSERT\n" +
+                        "SELECT o.entity_key, o.last_update_ts, o.feature1, o.feature2 " +
+                        "FROM OLDW o INNER JOIN NEWW n ON o.ENTITY_KEY=n.ENTITY_KEY AND o.LAST_UPDATE_TS > n.LAST_UPDATE_TS"
```

Comment:
We're upserting here again all the rows which were newer than the newly updated row w.r.t. `LAST_UPDATE_TS`, this triggers a corresponding update of history table rows, is this intentional? because I think we can directly update the history table rows in this trigger, right?

Reply:
This set of triggers was originally proposed in DB-12032 and serves as a baseline for other implementations.
The idea is that all new data goes to the feature table first and the history table is updated using the first two triggers upon every insert/update. The last trigger is needed to maintain the most recent data in the feature table.
Alternatively, as implemented by the last two tests, _updateHistoryTable_ and _updateHistoryTableConcurrently_, we may send all new data, not necessarily ordered, to the history table first and maintain the feature table so that it always contains the most recent data for each entity.
The goal is to compare the two approaches (in two modes: batch and concurrent) and hopefully show that the cleaner approach based on history table updates is also more performant. Currently, both approaches are not functioning as expected due to bugs revealed during this micro-benchmark development.

Reply:
Thanks for the thorough explanation, I understand it better now.